README.md
# a.el

Emacs Lisp functions for dealing with associative structures in a uniform and functional way.

Inspired by Clojure, dash, and seq.el.

These functions can take association lists, hash tables, and in some cases vectors (where the index is considered the key).

This library copies the names and semantics of the Clojure standard library. If you know Clojure then just add `a-` to the function name. The only exceptions are:

- `a-alist` is an association list contructor, it has no Clojure counterpart.
- `a-has-key?` is the equivalent of Clojure's `contains?`. This historical naming mistake won't be fixed in Clojure, but we can fix it here.
- predicates have both a `?` and a `-p` version, e.g. `a-has-key-p`. Use the latter if you want greater consistency with existing Elisp code.

All functions in this library are pure, they do not mutate their arguments.

## Alternatives

The one unique selling point for `a.el` is that it lets you reuse your Clojure experience in Emacs, so Clojure programmers can be productive quickly. If you are not experienced in Clojure, or you want to do things "the Emacs way", then consider these alternatives.

- [map.el](https://github.com/emacs-mirror/emacs/blob/master/lisp/emacs-lisp/map.el) comes bundled with recent Emacs versions
- [asoc.el](https://github.com/troyp/asoc.el) Association list library
- [ht.el](https://github.com/Wilfred/ht.el) Hash table library
- [kv.el](https://github.com/nicferrier/emacs-kv) A collection of tools for dealing with key/value data structures such as plists, alists and hash-tables.

## Requirements

a.el relies on features that are part of Emacs 25, so you need Emacs 25 or later. There are no other dependencies.

## Installation

a.el is available from Lambda Island ELPA. Add this to your `~/.emacs` or `~/.emacs.d/init.el`

``` emacs-lisp
(require 'package)
(add-to-list 'package-archives
             '("lambdaisland" . "http://lambdaisland.github.io/elpa/") t)
```

## Functions

``` emacs-lisp
(a-list :foo 5 :bar 6)
;;=> ((:foo . 5) (:bar . 6))

(setq m (a-list :foo 5 :bar 6))
(setq h (a-hash-table :abc 123 :def 456))

(a-associative? m)
;;=> t
(a-associative? h)
;;=> t

(a-get m :foo)
;;=> 5
(a-get h :abc)
;;=> 123

(a-assoc m :foo 7 :baq 20)
;;=> ((:baq . 20) (:foo . 7) (:bar . 6))
(a-assoc h :foo 7)
;;=> #s(hash-table ... (:abc 123 :def 456 :foo 7 ...))

(a-keys m)
;;=> (:foo :bar)
(a-keys h)
;;=> (:def :abc)

(a-vals m)
;;=> (5 6)
(a-vals h)
;;=> (456 123)

(a-equal m (a-list :bar 6 :foo 5))
;;=> t

(a-has-key? m :bar)
;;=> t

(a-count h)
;;=> 2

(a-dissoc m :foo)
;;=> ((:bar . 6))

(a-assoc-in (a-list :name "Arne")
            [:stats :score] 100)
;;=> ((:name . "Arne") (:stats . ((:score . 100))))

(a-merge m h (a-list :and :more))
;;=> ((:and . :more) (:abc . 123) (:def . 456) (:foo . 5) (:bar . 6))

(a-merge-with '+ m (a-list :foo 10))
;;=> ((:foo . 15) (:bar . 6))

(a-update (a-list :name "Arne") :name 'concat " Brasseur")
;;=> ((:name . "Arne Brasseur"))

(setq player (a-list :name "Arne" :stats (a-list :score 99)))
(a-update-in player  [:stats :score] '+ 1)
;;=> ((:name . "Arne") (:stats (:score . 100)))
```

## LICENSE

&copy; Arne Brasseur 2017

Distributed under the terms of the GNU General Public License, version 3.0 or later. See LICENSE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
(this space intentionally left blank)

<!-- ### Changed -->
<!-- ### Added -->
<!-- ### Changed -->
<!-- ### Removed -->
<!-- ### Fixed -->

## [0.1.0] - 2018-08-06
### Added

- Added `a-get*`

## [0.1.0alpha4] - 2017-07-20
### Fixed
- Fixed `a-get`, `a-has-key`, and `a-get-in` to correctly work with string keys.

## [0.1.0alpha3] - 2017-07-16
### Added
- `a-dissoc` Remove a key from a collection
- `a-associative?` / `a-associative-p` Returns true for association lists and hash tables
- `a-merge-with` Merge, resolving conflicts with a given function

### Changed
- `a-equal` now works on any value, and does a deep comparison
- `a-equal` terminates early
- Removed dash dependency

## [0.1.0alpha2] - 2017-07-15
### Added
- `a-hash-table` Hash table constructor, similar to `a-list`

[Unreleased]: https://github.com/plexus/a.el/compare/v0.1.0alpha3...HEAD
[0.1.0alpha4]: https://github.com/plexus/a.el/compare/v0.1.0alpha3...v0.1.0alpha4
[0.1.0alpha3]: https://github.com/plexus/a.el/compare/v0.1.0alpha2...v0.1.0alpha3
[0.1.0alpha2]: https://github.com/plexus/a.el/compare/6760b4edb7cf...v0.1.0alpha2


a-test.el
;;; a-test.el --- Associative function              -*- lexical-binding: t; -*-

;; Copyright (C) 2017  Arne Brasseur

;; Author: Arne Brasseur <arne@arnebrasseur.net>

;; This program is free software; you can redistribute it and/or modify it under
;; the terms of the Mozilla Public License Version 2.0

;; This program is distributed in the hope that it will be useful, but WITHOUT
;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
;; FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
;; details.

;; You should have received a copy of the Mozilla Public License along with this
;; program. If not, see <https://www.mozilla.org/media/MPL/2.0/index.txt>.

;;; Commentary:

;; Library for dealing with associative data structures: alists, hash-maps, and
;; vectors (for vectors, the indices are treated as keys)

;;; Code:

(require 'a)
(require 'ert)

(ert-deftest a-associative?-test ()
  (should (a-associative? nil))
  (should (a-associative? (a-list)))
  (should (a-associative? (a-list :foo 1)))
  (should (a-associative? (a-hash-table :foo 1)))
  (should (not (a-associative? '(1 2 3)))))

(ert-deftest a-get-test ()
  (should (equal (a-get '((:foo . 5)) :foo) 5))
  (should (equal (a-get '((:foo . 5)) :bar) nil))
  (should (equal (a-get '((:foo . 5)) :bar :fallback) :fallback))
  (should (equal (a-get [1 2 3] 1) 2))
  (should (equal (a-get [1 2 3] 5) nil))
  (should (equal (a-get [1 2 3] 5 :fallback) :fallback))
  (should-error (a-get 5 :nope))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 456 hash)
    (should (equal (a-get hash :bar) 456))
    (should (equal (a-get hash :baz) nil))
    (should (equal (a-get hash :baz :baq) :baq))))

(ert-deftest a-get-in-test ()
  (should (equal (a-get-in [1 2 [3 4 [5] 6]] [2 2 0]) 5))
  (should (equal (a-get-in [1 ((:a . ((:b . [3 4 5]))))] [1 :a :b 2]) 5))
  (should (equal (a-get-in [] []) []))
  (should (equal (a-get-in [] [2] :foo) :foo)))

(ert-deftest a-has-key-test ()
  (should (a-has-key [1 2 3] 2))
  (should (not (a-has-key [1 2 3] 3)))
  (should (not (a-has-key [1 2 3] -1)))
  (should (not (a-has-key [1 2 3] :foo)))
  (should-error (a-has-key? 1 :nope))

  (should (a-has-key '((:a . 5)) :a))
  (should (not (a-has-key '((:a . 5)) :b)))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (should (equal (a-has-key hash :foo) t))
    (should (equal (a-has-key hash :bar) nil))))

(ert-deftest a-assoc-test ()
  (should (equal (a-assoc '() :foo :bar) '((:foo . :bar))))
  (should (equal (a-assoc '((:foo . :baz)) :foo :bar) '((:foo . :bar))))
  (should (equal (a-assoc '((:foo . :baz))
                          :foo :bar
                          :baz :baq) '((:baz . :baq) (:foo . :bar))))

  (should (equal (a-assoc [1 2 3] 1 :foo) [1 :foo 3]))
  (should (equal (a-assoc [1 2 3] 5 :foo) [1 2 3 nil nil :foo]))
  (should-error (a-assoc '() :foo))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (should (a-equal (a-assoc hash :bar :val) '((:bar . :val) (:foo . 123))))))

(ert-deftest a-keys-test ()
  (should (equal (a-keys '((:a . 1) (:b . 2))) '(:a :b)))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 123 hash)

    (should (equal (a-keys hash) '(:bar :foo)))))

(ert-deftest a-vals-test ()
  (should (equal (a-vals '((:a . 1) (:b . 2))) '(1 2)))
  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 456 hash)
    (should (equal (a-vals hash) '(456 123)))))

(ert-deftest a-reduce-kv ()
  (should (equal
           (a-reduce-kv
            (lambda (acc k v)
              (cons
               (concat (symbol-name k) "--" (number-to-string v))
               acc))
            nil '((:a . 1) (:b . 2)))
           '(":b--2" ":a--1"))))

(ert-deftest a-reduce-kv-test ()
  (should (equal
           (a-reduce-kv
            (lambda (acc k v)
              (cons
               (concat (symbol-name k) "--" (number-to-string v))
               acc))
            nil '((:a . 1) (:b . 2)))
           '(":b--2" ":a--1"))))

(ert-deftest a-equal-test ()
  (should (a-equal '((:a . 1) (:b . 2)) '((:b . 2) (:a . 1))))
  (should (a-equal '((:a . 1) (:b . 2)) '((:a . 1) (:b . 2))))
  (should (not (a-equal '((:a . 1) (:c . 2)) '((:a . 1) (:b . 2)))))
  (should (not (a-equal '((:a . 1) (:b . 2)) '((:a . 1) (:b . 3)))))
  (should (not (a-equal '((:a . 1) (:b . 2)) '((:a . 1)))))
  (should (not (a-equal '((:a . 1) (:b . 2)) '((:a . 1) (:b . 2) (:c . 3)))))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 456 hash)
    (should (a-equal hash '((:foo . 123) (:bar . 456)))))

  (should (a-equal '((:a . 1) (:b . 2)) '((:b . 2) (:a . 1))))

  (should (a-equal 1 1))
  (should (a-equal '(1 2 3) [1 2 3]))
  (should (not (a-equal '(1 2 3 4) [1 2 3])))
  (should (not (a-equal '(1 2 3) [1 2 3 4])))
  (should (a-equal (a-list :foo (a-hash-table :bar [1 2] :baz nil))
                   (a-hash-table :foo (a-list :bar '(1 2) :baz nil)))))

(ert-deftest a-merge-test ()
  (should
   (a-equal
    (a-merge
     '((:a . 1) (:b . 2)))
    '((:b . 2) (:a . 1))))

  (should
   (a-equal
    (a-merge
     nil
     '((:a . 1) (:b . 2)))
    '((:b . 2) (:a . 1))))

  (should
   (a-equal
    (a-merge
     nil
     '((:a . 1) (:b . 2))
     '((:c . 3) (:b . 5)))
    '((:a . 1) (:b . 5) (:c . 3)))))

(ert-deftest a-merge-with-test ()
  (should (a-equal (a-merge-with '+ (a-list :x 5 :y 3) (a-list :x 7 :z 6))
                   (a-list :z 6 :x 12 :y 3))))

(ert-deftest a-alist ()
  (should
   (a-equal
    (a-alist :a 1 :b 2)
    '((:b . 2) (:a . 1)))))

(ert-deftest a-assoc-in-test ()
  (should
   (equal
    (a-assoc-in (a-alist :foo (a-alist :bar [1 2 3])) [:foo :bar 2] 5)
    '((:foo
       (:bar . [1 2 5])))))

  (should
   (equal
    (a-assoc-in (a-alist :foo nil) [:foo :bar 2] 5)
    '((:foo . ((:bar . ((2 . 5)))))))))

(ert-deftest a-dissoc-test ()
  (should (a-equal
           (a-dissoc (a-list :foo 1 :baz 2 :baq 3) :baz)
           (a-list :foo 1 :baq 3)))
  (should (a-equal
           (a-dissoc (a-list :foo 1 :baz 2 :baq 3) :baq :foo)
           (a-list :baz 2)))
  (should (a-equal
           (a-dissoc (a-hash-table :foo 1 :baz 2 :baq 3) :baq)
           (a-hash-table  :foo 1 :baz 2))))

(ert-deftest a-update-in-test ()
  (should
   (equal
    (a-update-in (a-alist :foo (a-alist :bar [1 2 "x"]))
                 [:foo :bar 2]
                 #'concat "y")
    '((:foo (:bar . [1 2 "xy"]))))))

(provide 'a-test)
;;; a-test.el ends here


test/a-test.el
;;; a-test.el --- Associative data structure functions: tests              -*- lexical-binding: t; -*-

;; Copyright (C) 2017  Arne Brasseur

;; Author: Arne Brasseur <arne@arnebrasseur.net>

;; This file is not part of GNU Emacs.

;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Test suite for a.el

;;; Code:

(require 'a)
(require 'ert)

(ert-deftest a-associative?-test ()
  (should (a-associative? nil))
  (should (a-associative? (a-list)))
  (should (a-associative? (a-list :foo 1)))
  (should (a-associative? (a-hash-table :foo 1)))
  (should (not (a-associative? '(1 2 3)))))

(ert-deftest a-get-test ()
  (should (equal (a-get '((:foo . 5)) :foo) 5))
  (should (equal (a-get '((:foo . 5)) :bar) nil))
  (should (equal (a-get '((:foo . 5)) :bar :fallback) :fallback))
  (should (equal (a-get [1 2 3] 1) 2))
  (should (equal (a-get [1 2 3] 5) nil))
  (should (equal (a-get [1 2 3] 5 :fallback) :fallback))
  (should-error (a-get 5 :nope))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 456 hash)
    (should (equal (a-get hash :bar) 456))
    (should (equal (a-get hash :baz) nil))
    (should (equal (a-get hash :baz :baq) :baq))))

(ert-deftest a-get-in-test ()
  (should (equal (a-get-in [1 2 [3 4 [5] 6]] [2 2 0]) 5))
  (should (equal (a-get-in [1 ((:a . ((:b . [3 4 5]))))] [1 :a :b 2]) 5))
  (should (equal (a-get-in [] []) []))
  (should (equal (a-get-in [] [2] :foo) :foo)))

(ert-deftest a-has-key-test ()
  (should (a-has-key [1 2 3] 2))
  (should (not (a-has-key [1 2 3] 3)))
  (should (not (a-has-key [1 2 3] -1)))
  (should (not (a-has-key [1 2 3] :foo)))
  (should-error (a-has-key? 1 :nope))

  (should (a-has-key '((:a . 5)) :a))
  (should (not (a-has-key '((:a . 5)) :b)))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (should (equal (a-has-key hash :foo) t))
    (should (equal (a-has-key hash :bar) nil))))

(ert-deftest a-assoc-test ()
  (should (equal (a-assoc '() :foo :bar) '((:foo . :bar))))
  (should (equal (a-assoc '((:foo . :baz)) :foo :bar) '((:foo . :bar))))
  (should (equal (a-assoc '((:foo . :baz))
                          :foo :bar
                          :baz :baq) '((:baz . :baq) (:foo . :bar))))

  (should (equal (a-assoc [1 2 3] 1 :foo) [1 :foo 3]))
  (should (equal (a-assoc [1 2 3] 5 :foo) [1 2 3 nil nil :foo]))
  (should-error (a-assoc '() :foo))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (should (a-equal (a-assoc hash :bar :val) '((:bar . :val) (:foo . 123))))))

(ert-deftest a-keys-test ()
  (should (equal (a-keys '((:a . 1) (:b . 2))) '(:a :b)))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 123 hash)

    (should (equal (sort (a-keys hash) #'string<) '(:bar :foo)))))

(ert-deftest a-vals-test ()
  (should (equal (a-vals '((:a . 1) (:b . 2))) '(1 2)))
  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 456 hash)
    (should (equal (sort (a-vals hash) #'<) '(123 456)))))

(ert-deftest a-reduce-kv ()
  (should (equal
           (a-reduce-kv
            (lambda (acc k v)
              (cons
               (concat (symbol-name k) "--" (number-to-string v))
               acc))
            nil '((:a . 1) (:b . 2)))
           '(":b--2" ":a--1"))))

(ert-deftest a-reduce-kv-test ()
  (should (equal
           (a-reduce-kv
            (lambda (acc k v)
              (cons
               (concat (symbol-name k) "--" (number-to-string v))
               acc))
            nil '((:a . 1) (:b . 2)))
           '(":b--2" ":a--1"))))

(ert-deftest a-equal-test ()
  (should (a-equal '((:a . 1) (:b . 2)) '((:b . 2) (:a . 1))))
  (should (a-equal '((:a . 1) (:b . 2)) '((:a . 1) (:b . 2))))
  (should (not (a-equal '((:a . 1) (:c . 2)) '((:a . 1) (:b . 2)))))
  (should (not (a-equal '((:a . 1) (:b . 2)) '((:a . 1) (:b . 3)))))
  (should (not (a-equal '((:a . 1) (:b . 2)) '((:a . 1)))))
  (should (not (a-equal '((:a . 1) (:b . 2)) '((:a . 1) (:b . 2) (:c . 3)))))

  (let ((hash (make-hash-table :test #'equal)))
    (puthash :foo 123 hash)
    (puthash :bar 456 hash)
    (should (a-equal hash '((:foo . 123) (:bar . 456)))))

  (should (a-equal '((:a . 1) (:b . 2)) '((:b . 2) (:a . 1))))

  (should (a-equal 1 1))
  (should (a-equal '(1 2 3) [1 2 3]))
  (should (not (a-equal '(1 2 3 4) [1 2 3])))
  (should (not (a-equal '(1 2 3) [1 2 3 4])))
  (should (a-equal (a-list :foo (a-hash-table :bar [1 2] :baz nil))
                   (a-hash-table :foo (a-list :bar '(1 2) :baz nil)))))

(ert-deftest a-merge-test ()
  (should
   (a-equal
    (a-merge
     '((:a . 1) (:b . 2)))
    '((:b . 2) (:a . 1))))

  (should
   (a-equal
    (a-merge
     nil
     '((:a . 1) (:b . 2)))
    '((:b . 2) (:a . 1))))

  (should
   (a-equal
    (a-merge
     nil
     '((:a . 1) (:b . 2))
     '((:c . 3) (:b . 5)))
    '((:a . 1) (:b . 5) (:c . 3)))))

(ert-deftest a-merge-with-test ()
  (should (a-equal (a-merge-with '+ (a-list :x 5 :y 3) (a-list :x 7 :z 6))
                   (a-list :z 6 :x 12 :y 3))))

(ert-deftest a-alist ()
  (should
   (a-equal
    (a-alist :a 1 :b 2)
    '((:b . 2) (:a . 1)))))

(ert-deftest a-assoc-in-test ()
  (should
   (equal
    (a-assoc-in (a-alist :foo (a-alist :bar [1 2 3])) [:foo :bar 2] 5)
    '((:foo
       (:bar . [1 2 5])))))

  (should
   (equal
    (a-assoc-in (a-alist :foo nil) [:foo :bar 2] 5)
    '((:foo . ((:bar . ((2 . 5)))))))))

(ert-deftest a-dissoc-test ()
  (should (a-equal
           (a-dissoc (a-list :foo 1 :baz 2 :baq 3) :baz)
           (a-list :foo 1 :baq 3)))
  (should (a-equal
           (a-dissoc (a-list :foo 1 :baz 2 :baq 3) :baq :foo)
           (a-list :baz 2)))
  (should (a-equal
           (a-dissoc (a-hash-table :foo 1 :baz 2 :baq 3) :baq)
           (a-hash-table  :foo 1 :baz 2))))

(ert-deftest a-update-in-test ()
  (should
   (equal
    (a-update-in (a-alist :foo (a-alist :bar [1 2 "x"]))
                 [:foo :bar 2]
                 #'concat "y")
    '((:foo (:bar . [1 2 "xy"]))))))

(ert-deftest a-get*-test ()
  (let ((alphabets (a-list "Greek" (a-list 1 (a-list 'letter "α"
                                                     'name "alpha")
                                           2 (a-list 'letter "β"
                                                     'name "beta"))
                           "English" (a-list 1 (a-list 'letter "a"
                                                       'name "A")
                                             2 (a-list 'letter "b"
                                                       'name "B")))))
    (should (equal (a-get* alphabets "Greek" 1 'letter)
                   "α"))))

;;; a-test.el ends here


test/test-helper.el
;;; test-helper.el --- Associative data structure functions: unit test helper functions -*- lexical-binding: t; -*-

;; Copyright (C) 2017  Arne Brasseur

;; Author: Arne Brasseur <arne@arnebrasseur.net>

;; This program is free software; you can redistribute it and/or modify it under
;; the terms of the Mozilla Public License Version 2.0

;; This program is distributed in the hope that it will be useful, but WITHOUT
;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
;; FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
;; details.

;; You should have received a copy of the Mozilla Public License along with this
;; program. If not, see <https://www.mozilla.org/media/MPL/2.0/index.txt>.

;;; Commentary:

;; Unit Test setup and helper functions for a.el

;;; Code:

(message "Running tests on Emacs %s" emacs-version)

(let* ((current-file (if load-in-progress load-file-name (buffer-file-name)))
       (source-directory (locate-dominating-file current-file "Cask"))
       ;; Do not load outdated byte code for tests
       (load-prefer-newer t))
  ;; Load the file under test
  (load (expand-file-name "a" source-directory)))

;; test-helper.el ends here


a.el
;;; a.el --- Associative data structure functions   -*- lexical-binding: t; -*-

;; Copyright (C) 2017  Arne Brasseur

;; Author: Arne Brasseur <arne@arnebrasseur.net>
;; URL: https://github.com/plexus/a.el
;; Keywords: lisp
;; Version: 0.1.0alpha4
;; Package-Requires: ((emacs "25"))

;; This file is not part of GNU Emacs.

;; This file is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 3, or (at your option)
;; any later version.

;; This file is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to
;; the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
;; Boston, MA 02110-1301, USA.

;;; Commentary:

;; Library for dealing with associative data structures: alists, hash-maps, and
;; vectors (for vectors, the indices are treated as keys).
;;
;; This library is largely inspired by Clojure, it has many of the functions
;; found in clojure.core, prefixed with `a-'. All functions treat their
;; arguments as immutable, so e.g. `a-assoc' will clone the hash-table or alist
;; it is given. Keep this in mind when writing performance sensitive code.

;;; Code:

(eval-when-compile (require 'subr-x)) ;; for things like hash-table-keys

(require 'cl-lib)
(require 'seq)

(defun a-associative-p (obj)
  (or (not obj)
      (hash-table-p obj)
      (and (consp obj) (consp (car obj)))))

(defalias 'a-associative? 'a-associative-p)

(defun a-get (map key &optional not-found)
  "Return the value MAP mapped to KEY, NOT-FOUND or nil if key not present."
  (cond
   ;; own implementation instead of alist-get so keys are checked with equal
   ;; instead of eq
   ((listp map)
    (a--alist-get map key not-found))

   ((vectorp map)
    (if (a-has-key? map key)
        (aref map key)
      not-found))

   ((hash-table-p map)
    (gethash key map not-found))
   (t (user-error "Not associative: %S" map))))

(defun a--alist-get (map key &optional not-found)
  "Like alist-get, but uses equal instead of eq to look up in map MAP key KEY.
Returns NOT-FOUND if the key is not present, or `nil' if
NOT-FOUND is not specified."
  (cl-block nil
    (seq-doseq (pair map)
      (when (equal (car pair) key)
        (cl-return (cdr pair))))
    not-found))

(defun a-get-in (m ks &optional not-found)
  "Look up a value in a nested associative structure.

Given a data structure M, and a sequence of keys KS, find the
value found by using each key in turn to do a lookup in the next
\"layer\". Return `nil' if the key is not present, or the NOT-FOUND
value if supplied."
  (let ((result m))
    (cl-block nil
      (seq-doseq (k ks)
        (if (a-has-key? result k)
            (setq result (a-get result k))
          (cl-return not-found)))
      result)))

(defmacro a-get* (&rest keys)
  "Look up a value in a nested associative structure.

Like a-get-in, but takes the key sequence KEYS directly as vararg
arguments, rather than as a single sequence."
  (cl-labels ((rec (keys)
                   `(a-get ,(if (and (consp (cdr keys))
                                     (cddr keys))
                                (rec (cdr keys))
                              (cadr keys))
                           ,(car keys))))
    (rec (nreverse keys))))

(defun a-has-key (coll k)
  "Check if the given associative collection COLL has a certain key K."
  (cond
   ((listp coll)         (not (eq (a--alist-get coll k :not-found) :not-found)))
   ((vectorp coll)       (and (integerp k) (< -1 k (length coll))))
   ((hash-table-p coll)  (not (eq (gethash k coll :not-found) :not-found)))
   (t (user-error "Not associative: %S" coll))))

(defalias 'a-has-key? 'a-has-key)

(defun a-assoc-1 (coll k v)
  "Like `a-assoc', (in COLL assoc K with V) but only takes a single k-v pair.
Internal helper function."
  (cond
   ((listp coll)
    (if (a-has-key? coll k)
        (mapcar (lambda (entry)
                  (if (equal (car entry) k)
                      (cons k v)
                    entry))
                coll)
      (cons (cons k v) coll)))

   ((vectorp coll)
    (if (and (integerp k) (> k 0))
        (if (< k (length coll))
            (let ((copy (copy-sequence coll)))
              (aset copy k v)
              copy)
          (vconcat coll (make-list (- k (length coll)) nil) (list v)))))

   ((hash-table-p coll)
    (let ((copy (copy-hash-table coll)))
      (puthash k v copy)
      copy))))

(defun a-assoc (coll &rest kvs)
  "Return an updated collection COLL, associating values with keys KVS."
  (when (not (cl-evenp (a-count kvs)))
    (user-error "a-assoc requires an even number of arguments!"))
  (seq-reduce (lambda (coll kv)
                  (seq-let [k v] kv
                    (a-assoc-1 coll k v)))
              (seq-partition kvs 2)
              coll))

(defun a-keys (coll)
  "Return the keys in the collection COLL."
  (cond
   ((listp coll)
    (mapcar #'car coll))

   ((hash-table-p coll)
    (hash-table-keys coll))))

(defun a-vals (coll)
  "Return the values in the collection COLL."
  (cond
   ((listp coll)
    (mapcar #'cdr coll))

   ((hash-table-p coll)
    (hash-table-values coll))))

(defun a-reduce-kv (fn from coll)
  "Reduce with FN starting from FROM the collection COLL.
Reduce an associative collection COLL, starting with an initial
value of FROM. The reducing function FN receives the intermediate
value, key, and value."
  (seq-reduce (lambda (acc key)
                  (funcall fn acc key (a-get coll key)))
              (a-keys coll)
              from))

(defun a-count (coll)
  "Count the number of key-value pairs in COLL.
Like length, but can also return the length of hash tables."
  (cond
   ((seqp coll)
    (length coll))

   ((hash-table-p coll)
    (hash-table-count coll))))

(defun a-equal (a b)
  "Compare collections A, B for value equality.

Associative collections (hash tables and a-lists) are considered
equal if they contain equal key-value pairs, regardless of order.

Sequences (lists or vectors) are considered equal if they contain
the same elements in the same order.

Collection elements are compared using `a-equal'. In other words,
the equality check is recursive, resulting in a \"deep\" equality
check.

Anything that isn't associative or a sequence is compared with
`equal'."
  (cond
   ((and (a-associative? a) (a-associative? b))
    (or (equal a b)
        (when (eq (a-count a) (a-count b))
          (cl-block nil
            (seq-doseq (k (a-keys a))
              (when (not (a-equal (a-get a k) (a-get b k)))
                (cl-return nil)))
            t))))
   ((and (sequencep a) (sequencep b))
    (and (eq (length a) (length b))
         (or (and (seq-empty-p a) (seq-empty-p b))
             (and (a-equal (elt a 0) (elt b 0))
                  (a-equal (seq-drop a 1) (seq-drop b 1))))))
   (t
    (equal a b))))

(defalias 'a-equal? 'a-equal)

(defun a-merge (&rest colls)
  "Merge multiple associative collections.
Return the type of the first collection COLLS."
  (seq-reduce (lambda (this that)
                (a-reduce-kv (lambda (coll k v)
                               (a-assoc coll k v))
                             this
                             that))
              (cdr colls)
              (car colls)))

(defun a-merge-with (f &rest colls)
  "Merge multiple associative collections.
Return the type of the first collection COLLS. If a key exists in
both, then combine the associated values by calling f on them."
  (seq-reduce (lambda (this that)
                (a-reduce-kv (lambda (coll k v)
                               (a-assoc coll k (if (a-has-key coll k)
                                                   (funcall f v (a-get coll k))
                                                 v)))
                             this
                             that))
              (cdr colls)
              (car colls)))

(defun a-alist (&rest kvs)
  "Create an association list from the given keys and values KVS.
Arguments are simply provided in sequence, rather than as lists or cons cells.
For example: (a-alist :foo 123 :bar 456)"
  (mapcar (lambda (kv) (cons (car kv) (cadr kv))) (seq-partition kvs 2)))

(defalias 'a-list 'a-alist)

(defun a-hash-table (&rest kvs)
  "Create a hash table from the given keys and values KVS.
Arguments are simply provided in sequence, rather than as lists
or cons cells. As \"test\" for the hash table, equal is used. The
hash table is created without extra storage space, so with a size
equal to amount of key-value pairs, since it is assumed to be
treated as immutable.
For example: (a-hash-table :foo 123 :bar 456)"
  (let* ((kv-pairs (seq-partition kvs 2))
         (hash-map (make-hash-table :test 'equal :size (length kv-pairs))))
    (seq-do (lambda (pair)
              (puthash (car pair) (cadr pair) hash-map))
            kv-pairs)
    hash-map))

(defun a-assoc-in (coll keys value)
  "In collection COLL, at location KEYS, associate value VALUE.
Associates a value in a nested associative collection COLL, where
KEYS is a sequence of keys and VALUE is the new value and returns
a new nested structure. If any levels do not exist, association
lists will be created."
  (cl-case (length keys)
    (0 coll)
    (1 (a-assoc-1 coll (elt keys 0) value))
    (t (a-assoc-1 coll
                  (elt keys 0)
                  (a-assoc-in (a-get coll (elt keys 0))
                              (seq-drop keys 1)
                              value)))))

(defun a-dissoc--list (list keys)
  "Return updated LIST with KEYS removed.
Internal helper. Use `a-dissoc' instead."
  (a-reduce-kv (lambda (res k v)
                 (if (member k keys)
                     res
                   (cons (cons k v) res)))
               nil
               list))

(defun a-dissoc--hash-table (table keys)
  "Return updated TABLE with KEYS removed.
Internal helper. Use `a-dissoc' instead."
  (let ((new-table (make-hash-table :size (hash-table-count table)
                                    :test (hash-table-test table)))
        (rest-keys (seq-remove (lambda (k)
                                 (member k keys))
                               (a-keys table))))
    (seq-doseq (k rest-keys)
      (puthash k (gethash k table) new-table))
    new-table))

(defun a-dissoc (coll &rest keys)
  "Return an updated version of collection COLL with the KEY removed."
  (cond
   ((listp coll) (a-dissoc--list coll keys))
   ((hash-table-p coll) (a-dissoc--hash-table coll keys))))

(defun a-update (coll key fn &rest args)
  "In collection COLL, at location KEY, apply FN with extra args ARGS.
'Updates' a value in an associative collection COLL, where KEY is
a key and FN is a function that will take the old value and any
supplied args and return the new value, and returns a new
structure. If the key does not exist, nil is passed as the old
value."
  (a-assoc-1 coll
             key
             (apply #'funcall fn (a-get coll key) args)))

(defun a-update-in (coll keys fn &rest args)
  "In collection COLL, at location KEYS, apply FN with extra args ARGS.
'Updates' a value in a nested associative collection COLL, where
KEYS is a sequence of keys and FN is a function that will take
the old value and any supplied ARGS and return the new value, and
returns a new nested structure. If any levels do not exist,
association lists will be created."
  (cl-case (length keys)
    (0 coll)
    (1 (apply #'a-update coll (elt keys 0) fn args))
    (t (a-assoc-1 coll
                  (elt keys 0)
                  (apply #'a-update-in
                         (a-get coll (elt keys 0))
                         (seq-drop keys 1)
                         fn
                         args)))))

(provide 'a)
;;; a.el ends here
